src/charm.py

```diff
@@ -6,38 +6,197 @@
+        self.unit.status = ActiveStatus()
+
+    def _construct_waypoint(self):
+        gateway = IstioWaypointResource(
+            metadata=Metadata(
+                name=f"{self.app.name}-{self.model.name}-waypoint",
+                namespace=self.model.name,
+                labels={"istio.io/waypoint-for": "service"},
+            ),
+            spec=IstioWaypointSpec(
+                gatewayClassName="istio-waypoint",
+                listeners=[
+                    Listener(
+                        name="mesh",
+                        port=15008,
+                        protocol="HBONE",
+                        allowedRoutes=AllowedRoutes(namespaces={"from": "All"}),
+                    )
+                ],
+            ),
+        )
+        gateway_resource = RESOURCE_TYPES["Gateway"]
+        return gateway_resource(
+            metadata=ObjectMeta.from_dict(gateway.metadata.dict()),
+            spec=gateway.spec.dict(),
+        )
+
+    def _sync_waypoint_resources(self):
+        resources_list = []
+        if not self.unit.is_leader():
+            raise RuntimeError("Waypoint can only be provided on the leader unit.")
+        krm = self._get_waypoint_resource_manager()
+        resource_to_append = self._construct_waypoint()
+        resources_list.append(resource_to_append)
+        krm.reconcile(resources_list)
+
+        if self.config["model-on-mesh"]:
+            self._update_labels()
+        else:
+            self._update_labels(False)
+
+    def _update_labels(self, add_labels=True):
+        """Add or remove specific labels from a namespace based on the add_labels flag.
+
+        Args:
+            add_labels (bool): If True, add labels; if False, remove labels.
         """
-        for relation in self.model.relations["service-mesh"]:
-            logger.error(relation.data[relation.app])
-            # Update the mesh
-        self.unit.status = ops.ActiveStatus()
+        try:
+            namespace = self.lightkube_client.get(Namespace, self.model.name)
+        except ApiError as e:
+            logger.error(f"Error checking namespace labels {e}")
+            return
+
+        if add_labels:
+            labels_to_add = {
+                "istio.io/use-waypoint": f"{self.app.name}-{self.model.name}-waypoint",
+                "istio.io/dataplane-mode": "ambient",
+            }
+            if namespace.metadata and namespace.metadata.labels:
+                namespace.metadata.labels.update(labels_to_add)
+        else:
+            labels_to_remove = {"istio.io/use-waypoint": None, "istio.io/dataplane-mode": None}
+
+            if (
+                namespace.metadata
+                and namespace.metadata.labels
+                and (
+                    "istio.io/use-waypoint" in namespace.metadata.labels
+                    or "istio.io/dataplane-mode" in namespace.metadata.labels
+                )
+            ):
+                namespace.metadata.labels.update(labels_to_remove)
```

Comment:
we can simplify this a little by (I left out the error checking as I'd probably make a mistake here, but you'd have to add it back):
```suggestion
        labels = {
            "istio.io/use-waypoint": None,
            "istio.io/dataplane-mode": None,
        }
        if add_labels:
            labels["istio.io/use-waypoint"] = f"{self.app.name}-{self.model.name}-waypoint"
            labels["istio.io/dataplane-mode"] = "ambient"
        namespace.metadata.labels.update(labels)
```